group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java

```diff
@@ -0,0 +1,2080 @@
+                    break;
+
+                case ConsumerGroupPartitionMetadataKey.HIGHEST_SUPPORTED_VERSION:
+                    groupMetadataManager.replay(
+                        (ConsumerGroupPartitionMetadataKey) key.message(),
+                        (ConsumerGroupPartitionMetadataValue) messageOrNull(value)
+                    );
+                    break;
+
+                case ConsumerGroupTargetAssignmentMemberKey.HIGHEST_SUPPORTED_VERSION:
+                    groupMetadataManager.replay(
+                        (ConsumerGroupTargetAssignmentMemberKey) key.message(),
+                        (ConsumerGroupTargetAssignmentMemberValue) messageOrNull(value)
+                    );
+                    break;
+
+                case ConsumerGroupTargetAssignmentMetadataKey.HIGHEST_SUPPORTED_VERSION:
+                    groupMetadataManager.replay(
+                        (ConsumerGroupTargetAssignmentMetadataKey) key.message(),
+                        (ConsumerGroupTargetAssignmentMetadataValue) messageOrNull(value)
+                    );
+                    break;
+
+                case ConsumerGroupCurrentMemberAssignmentKey.HIGHEST_SUPPORTED_VERSION:
+                    groupMetadataManager.replay(
+                        (ConsumerGroupCurrentMemberAssignmentKey) key.message(),
+                        (ConsumerGroupCurrentMemberAssignmentValue) messageOrNull(value)
+                    );
+                    break;
+
+                default:
+                    throw new IllegalStateException("Received an unknown record type " + key.version()
+                        + " in " + record);
+            }
+
+            lastWrittenOffset++;
+        }
+    }
+
+    @Test
+    public void testConsumerHeartbeatRequestValidation() {
+        MockPartitionAssignor assignor = new MockPartitionAssignor("range");
+        GroupMetadataManagerTestContext context = new GroupMetadataManagerTestContext.Builder()
+            .withAssignors(Collections.singletonList(assignor))
+            .build();
+        Exception ex;
+
+        ex = assertThrows(InvalidRequestException.class, () -> context.consumerGroupHeartbeat(
+            new ConsumerGroupHeartbeatRequestData()));
+        assertEquals("GroupId can't be empty.", ex.getMessage());
+
+        ex = assertThrows(InvalidRequestException.class, () -> context.consumerGroupHeartbeat(
+            new ConsumerGroupHeartbeatRequestData()
+                .setGroupId("foo")
+                .setMemberEpoch(0)));
+        assertEquals("RebalanceTimeoutMs must be provided in first request.", ex.getMessage());
+
+        ex = assertThrows(InvalidRequestException.class, () -> context.consumerGroupHeartbeat(
+            new ConsumerGroupHeartbeatRequestData()
+                .setGroupId("foo")
+                .setMemberEpoch(0)
+                .setRebalanceTimeoutMs(5000)));
+        assertEquals("TopicPartitions must be empty when (re-)joining.", ex.getMessage());
```

Comment:
Are the topic partitions not empty here?

Reply:
I see we check for null or !empty.

Reply:
Ok so null is not ok -- it must be explicitly empty.

Reply:
Right.